File: Algorithms/Sorting/JavaScript/Merge.js
/*
    Merge Sort
    Best Case: O(n log n)
    Worst Case: O(n log n)
*/
function merge(A, p, q, r) {
  var n1 = q - p + 1;
  var n2 = r - q;

  var L = [];
  var R = [];

  for(var i = 0; i < n1; i++) {
    L.push(A[p + i]);
  }

  for(var i = 0; i < n2; i++) {
    R.push(A[q + i + 1]);
  }

  var i = 0;
  var j = 0;
  var k = p;

  while(i < n1 && j < n2) {
    if(L[i] <= R[j]) {
      A[k] = L[i];
      i++;
    } else {
      A[k] = R[j];
      j++;
    }

    k++;
  }

  while(i < n1) {
    A[k] = L[i];
    i++;
    k++;
  }

  while(j < n2) {
    A[k] = R[j];
    j++;
    k++;
  }

  return A;
}

function mergeSort(A, p, r) {
  var B = A;
  
  if(p < r) {
    var q = (p + r) / 2;
    B = mergeSort(B, p, q);
    B = mergeSort(B, q + 1, r);
    B = merge(B, p, q, r);
  }
  
  return B;
}

console.log(mergeSort([2, 4, 5, 7, 1, 2, 3, 6], 0, 7));

File: Algorithms/Sorting/Java/Merge.java
package Algorithms.Sorting.Java;
import java.util.Arrays;

class Solution {
    public static void main(String[] args) {
        int[] arr = {2, 4, 5, 7, 1, 2, 3, 6};

        mergeSort(arr, 0, arr.length - 1);

        System.out.println(Arrays.toString(arr));
    }

    public static void merge(int[] A, int start, int middle, int end) {
        int leftBound = middle - start + 1;
        int rightBound = end - middle;

        int[] L = new int[leftBound];
        int[] R = new int[rightBound];

        for(int i = 0; i < leftBound; i++) {
            L[i] = A[start + i];
        }

        for(int i = 0; i < rightBound; i++) {
            R[i] = A[middle + i + 1];
        }

        int i = 0;
        int j = 0;
        int k = start;

        while(i < leftBound && j < rightBound) {
            if(L[i] <= R[j]) {
                A[k] = L[i];
                i++;
            } else {
                A[k] = R[j];
                j++;
            }
            k++;
        }

        while(i < leftBound) {
            A[k] = L[i];
            i++;
            k++;
        }

        while(j < rightBound) {
            A[k] = R[j];
            j++;
            k++;
        }
    }

    public static void mergeSort(int[] A, int start, int end) {
        if(start < end) {
            int middle = (int) Math.floor((start + end) / 2);
            mergeSort(A, start, middle);
            mergeSort(A, middle + 1, end);
            merge(A, start, middle, end);
        }
    }
}